File: src/main/java/com/mercateo/jsonschema/property/UnwrappedPropertyBuilder.java
package com.mercateo.jsonschema.property;

import com.fasterxml.jackson.annotation.JsonUnwrapped;
import com.mercateo.jsonschema.generictype.GenericType;
import javaslang.collection.HashMap;
import javaslang.collection.HashSet;
import javaslang.collection.List;
import javaslang.collection.Set;

import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class UnwrappedPropertyBuilder {

    private final PropertyBuilder propertyBuilder;

    private final Set<Class<? extends Annotation>> unwrapAnnotations;

    private ConcurrentHashMap<GenericType, Property> unwrappedProperties;

    private ConcurrentHashMap<GenericType, PropertyDescriptor> unwrappedPropertyDescriptors;

    public UnwrappedPropertyBuilder(PropertyBuilder propertyBuilder, Class<? extends Annotation>... unwrapAnnotations) {
        this.propertyBuilder = propertyBuilder;
        this.unwrapAnnotations = HashSet.of(unwrapAnnotations);

        this.unwrappedProperties = new ConcurrentHashMap<>();
        this.unwrappedPropertyDescriptors = new ConcurrentHashMap<>();
    }

    public Property from(Class<?> clazz) {
        return from(GenericType.of(clazz));
    }

    public Property from(GenericType<?> genericType) {
        return unwrappedProperties.computeIfAbsent(genericType, this::unwrap);
    }

    private Property unwrap(GenericType<?> genericType) {
        return unwrap(propertyBuilder.from(genericType));
    }

    private Property unwrap(Property property) {
        final PropertyDescriptor propertyDescriptor = unwrappedPropertyDescriptors.computeIfAbsent(
                property.genericType(), t -> createUnwrappedDescriptor(property));

        return ImmutableProperty.of(property.name(), propertyDescriptor, property.valueAccessor(),
                property.annotations());
    }

    private PropertyDescriptor createUnwrappedDescriptor(Property property) {
        List<Property> children = addChildren(List.empty(), property);
        return ImmutablePropertyDescriptor.of(property.genericType(), children, property
                .propertyDescriptor().annotations());
    }

    private List<Property> addChildren(List<Property> children, Property property) {
        return property.children()
                .map(this::unwrap)
                .flatMap(child -> {
                    if (!unwrapAnnotations.intersect(child.annotations().keySet()).isEmpty()) {
                        child = updateValueAccessors(child);
                        return addChildren(children, child);
                    } else {
                        return children.append(child);
                    }

        });
    }

    private Property updateValueAccessors(Property child) {
        final List<Property> children = child.children().map(c -> (Property) ImmutableProperty.of(c.name(), c.propertyDescriptor(),
                object -> {
                    Object intermediateObject = child.valueAccessor().apply(object);
                    return intermediateObject != null ? c.valueAccessor().apply(intermediateObject) : null;
                }, c.annotations())).toList();

        final PropertyDescriptor propertyDescriptor = child.propertyDescriptor();
        final PropertyDescriptor updatedPropertyDescriptor = ImmutablePropertyDescriptor.of(
                propertyDescriptor.genericType(),
                children,
                propertyDescriptor.annotations());
        return ImmutableProperty.of(child.name(), updatedPropertyDescriptor, child.valueAccessor(), child.annotations());
    }

}

File: src/main/java/com/mercateo/jsonschema/property/UnwrappedPropertyBuilder.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.generictype.GenericType
import java.util.concurrent.ConcurrentHashMap

class UnwrappedPropertyBuilder
@SafeVarargs
constructor(private val propertyBuilder: PropertyBuilder, vararg unwrapAnnotations: Class<out Annotation>) : PropertyBuilder {

    private val unwrapAnnotations: Set<Class<out Annotation>>

    private val unwrappedProperties: ConcurrentHashMap<GenericType<*>, Property>

    private val unwrappedPropertyDescriptors: ConcurrentHashMap<GenericType<*>, PropertyDescriptor>

    init {
        this.unwrapAnnotations = setOf(*unwrapAnnotations)

        this.unwrappedProperties = ConcurrentHashMap<GenericType<*>, Property>()
        this.unwrappedPropertyDescriptors = ConcurrentHashMap<GenericType<*>, PropertyDescriptor>()
    }

    override fun from(clazz: Class<*>): Property {
        return from(GenericType.of(clazz))
    }

    override fun from(genericType: GenericType<*>): Property {
        return unwrappedProperties.computeIfAbsent(genericType, { unwrap(it) })
    }

    private fun unwrap(genericType: GenericType<*>): Property {
        val addedUnwrappedPropertyDescriptors = mutableMapOf<GenericType<*>, PropertyDescriptor>()
        val property = unwrap(propertyBuilder.from(genericType), addedUnwrappedPropertyDescriptors)
        unwrappedPropertyDescriptors.putAll(addedUnwrappedPropertyDescriptors)
        return property
    }

    private fun unwrap(
            property: Property,
            addedUnwrappedProperties: MutableMap<GenericType<*>, PropertyDescriptor>
    ): Property {
        val genericType = property.genericType

        val propertyDescriptor =
                if (unwrappedPropertyDescriptors.containsKey(genericType)) {
                    unwrappedPropertyDescriptors[genericType]!!
                } else {
                    if (addedUnwrappedProperties.containsKey(genericType)) {
                        addedUnwrappedProperties[genericType]!!
                    } else {
                        createDescriptor(property, addedUnwrappedProperties)
                    }
                }
        return Property(property.name, propertyDescriptor, property.valueAccessor, property.annotations)
    }

    private fun createDescriptor(
            property: Property,
            addedUnwrappedProperties: MutableMap<GenericType<*>, PropertyDescriptor>
    ): PropertyDescriptor {
        val children = getChildren(property, addedUnwrappedProperties)
        val propertyType = property.propertyType
        val genericType = property.genericType
        val propertyDescriptor = PropertyDescriptorDefault(propertyType, genericType, children, property.propertyDescriptor.annotations)
        addedUnwrappedProperties.put(genericType, propertyDescriptor)
        return propertyDescriptor
    }

    private fun getChildren(
            property: Property,
            addedUnwrappedPropertyDescriptors: MutableMap<GenericType<*>, PropertyDescriptor>
    ): List<Property> {
        return property.children
                .map {
                    unwrap(it, addedUnwrappedPropertyDescriptors)
                }
                .flatMap {
                    val doUnwrapChild = !unwrapAnnotations.intersect(it.annotations.keys).isEmpty()
                    if (doUnwrapChild) {
                        getChildren(updateValueAccessors(it), addedUnwrappedPropertyDescriptors)
                    } else {
                        listOf(it)
                    }
                }
    }

    private fun updateValueAccessors(child: Property): Property {
        val children = child.children.map {
            Property(it.name, it.propertyDescriptor,
                    { instance ->
                        val intermediateObject = child.valueAccessor(instance)
                        if (intermediateObject != null) it.valueAccessor(intermediateObject) else null
                    }, it.annotations)
        }

        val propertyDescriptor = child.propertyDescriptor
        val updatedPropertyDescriptor = PropertyDescriptorDefault(
                propertyDescriptor.propertyType,
                propertyDescriptor.genericType,
                children,
                propertyDescriptor.annotations)
        return Property(child.name, updatedPropertyDescriptor, child.valueAccessor, child.annotations)
    }

}
